### PR TITLE
server: balance according to region count

### LINKED ISSUE #444: Balance region count according to store capacity
The goal of the current storage balance algorithm is to balance storage size between stores, not region count. However, storage size changes very dynamically, but the balance effect takes time to feedback.

For example, if a store has the biggest storage size in the cluster, the balance algorithm will try to remove peer from the store. Unfortunately, the current implementation to remove peer may increase the storage size temporarily, which will cause the balance algorithm keeping removing peer from the store.

If the size of different regions are relative equal, we can try to balance the region count according to store capacity instead. 
For example, we have store 1 with 1GB capacity, store 2 with 2GB capacity, and 6 regions, we can schedule 2 regions to store 1 and 4 regions to store 2. In this way, we will be not affected by the dynamically changing storage size.

### PR DESCRIPTION
**Pull request by huachaohuang**
We will adjust the balance speed and threshold according to the leader count and
region count, so that we can remove some dumb configurations.
Close #444 .

### CODE REVIEW

File: server/balancer.go [hunk 1]
```diff
@@ -22,19 +23,33 @@ import (
 
 var storeCacheInterval = 30 * time.Second
 
+func minBalanceDiff(count uint64) float64 {
```
Comment (siddontang) [inline server/balancer.go]: can you give some explanation for the diff and speed adjustment?

Comment (siddontang) [inline server/balancer.go]: add test to cover these functions

File: server/util.go [hunk 2]
```diff
@@ -350,3 +350,13 @@ func GetPDMembers(etcdClient *clientv3.Client) ([]*pdpb.PDMember, error) {
 
 	return members, nil
 }
+
+func roundUint64(value, min, max uint64) uint64 {
```
Comment (siddontang) [inline server/util.go]: add a test to cover this.
Btw, round is proper here?

File: server/store.go [hunk 3]
```diff
@@ -66,35 +66,52 @@ func (s *storeInfo) downTime() time.Duration {
 	return time.Since(s.stats.LastHeartbeatTS)
 }
 
+func (s *storeInfo) leaderCount() uint64 {
+	return uint64(s.stats.LeaderCount)
+}
+
 func (s *storeInfo) leaderRatio() float64 {
-	if s.stats.TotalRegionCount == 0 {
+	return float64(s.stats.LeaderCount)
```
Comment (siddontang) [inline server/store.go]: is it right here to return leader counter?

Comment (disksing) [inline server/store.go]: +1, it's strange to call it _ratio_ while it is not actually a ratio.

File: server/balancer.go [hunk 4]
```diff
@@ -22,19 +23,33 @@ import (
...
+	return math.Pow(2, math.Log10(float64(count)))
+}
+
+func adjustBalanceSpeed(sourceCount uint64, sourceRatio, targetRatio float64) (uint64, bool) {
```
Comment (disksing) [inline server/balancer.go]: Please add some comments for it.

File: server/cache.go [hunk 5]
```diff
@@ -460,9 +460,8 @@ func (c *clusterInfo) handleStoreHeartbeat(stats *pdpb.StoreStats) error {
 	}
 
 	store.stats.StoreStats = proto.Clone(stats).(*pdpb.StoreStats)
+	store.stats.LeaderCount = c.regions.getStoreLeaderCount(storeID)
 	store.stats.LastHeartbeatTS = time.Now()
-	store.stats.TotalRegionCount = c.regions.getRegionCount()
```
Comment (disksing) [inline server/cache.go]: Why remove it?

File: server/util.go (hunk 2 above)
Comment (disksing) [inline server/util.go]: I prefer _ensureRange_ or _restrictRange_.

File: server/cache.go (hunk 5 above)
Comment (huachaohuang) [inline server/cache.go]: We don't need to use that information in the store any more.

File: server/store.go (hunk 3 above)
Comment (huachaohuang) [inline server/store.go]: We don't have any other factors to calculate for now, so I just return the leader count.
How about rename to `leaderScore`?

Comment (disksing) [inline server/store.go]: I'm okay with `leaderScore`.

Comment (siddontang) [inline server/store.go]: +1

File: server/coordinator.go [hunk 6]
```diff
@@ -35,54 +35,60 @@ type coordinator struct {
...
+	operators  map[uint64]Operator
+	schedulers map[string]*scheduleController
 
 	histories *lruCache
 	events    *fifoCache
 }
 
 func newCoordinator(cluster *clusterInfo, opt *scheduleOption) *coordinator {
-	c := &coordinator{
+	ctx, cancel := context.WithCancel(context.Background())
```
Comment (overvenus) [inline server/coordinator.go]: Maybe it is better to use a context from outside?

Comment (siddontang) [inline server/coordinator.go]: the context is the root context for the coordinator itself, I think it is ok here.

File: server/balancer.go [hunk 7]
```diff
@@ -22,8 +23,30 @@ import (
 
 var storeCacheInterval = 30 * time.Second
 
+// minBalanceDiff returns the minimal diff to do balance. The formula is based
+// on experience to let the diff increase alone with the count slowly.
+func minBalanceDiff(count uint64) float64 {
+	return math.Pow(2, math.Log10(float64(count)))
```
Comment (siddontang) [inline server/balancer.go]: The algorithm has any reference?

File: server/balancer.go [hunk 8]
```diff
@@ -22,8 +23,30 @@ import (
...
+// diffCount, so we will not over-scheduled. Finally, the min balance diff
+// provides a buffer to make the cluster stable, so we don't need to schedule
+// very frequently.
+func adjustBalanceSpeed(sourceCount uint64, sourceScore, targetScore float64) (uint64, bool) {
+	if targetScore >= sourceScore {
+		return 0, false
+	}
+	diffRatio := 1 - targetScore/sourceScore
+	diffCount := diffRatio * float64(sourceCount)
+	return uint64(diffCount) / 4, diffCount > minBalanceDiff(sourceCount)
```
Comment (siddontang) [inline server/balancer.go]: why `/ 4` here?

File: server/balancer.go [hunk 9]
```diff
@@ -22,8 +23,30 @@ import (
...
+// First, we calculate the diffCount (the number of schedules to balance the
+// source and the target). Then, we adjust the schedule limit to less than the
+// diffCount, so we will not over-scheduled. Finally, the min balance diff
+// provides a buffer to make the cluster stable, so we don't need to schedule
+// very frequently.
+func adjustBalanceSpeed(sourceCount uint64, sourceScore, targetScore float64) (uint64, bool) {
+	if targetScore >= sourceScore {
+		return 0, false
+	}
+	diffRatio := 1 - targetScore/sourceScore
```
Comment (siddontang) [inline server/balancer.go]: sourceScore can be 0?

Comment (huachaohuang) [inline server/balancer.go]: It can't be 0, since `sourceScore > targetScore`, and `targetScore >= 0`.

File: server/balancer.go (hunk 8 above)
Comment (huachaohuang) [inline server/balancer.go]: Well, if we want to balance two store at once, we can use `diffCount / 2` here, but consider that the store information may not up to date and the store feedback may delay, a smaller limit can help to reduce the effect of stale information. I choose `diffCount / 4` according to my experience, we may need to add a config to experiment it.

Comment (siddontang) [inline server/balancer.go]: I prefer config later.

File: server/balancer.go (hunk 7 above)
Comment (disksing) [inline server/balancer.go]: It will limit diff to 32 for 100k regions, seems too small.
How about `sqrt(count)`?

File: server/api/store.go [hunk 10]
```diff
@@ -36,26 +36,24 @@ type storeStatus struct {
...
 	RegionCount        uint32            `json:"region_count"`
 	SendingSnapCount   uint32            `json:"sending_snap_count"`
 	ReceivingSnapCount uint32            `json:"receiving_snap_count"`
-	StartTime          time.Time         `json:"start_time"`
 	ApplyingSnapCount  uint32            `json:"applying_snap_count"`
 	IsBusy             bool              `json:"is_busy"`
 
-	StartTS          time.Time         `json:"start_ts"`
-	LastHeartbeatTS  time.Time         `json:"last_heartbeat_ts"`
-	TotalRegionCount int               `json:"total_region_count"`
```
Comment (siddontang) [inline server/api/store.go]: no need TotalRegionCount now?

File: server/balancer_test.go [hunk 11]
```diff
@@ -116,16 +112,74 @@ func (c *testClusterInfo) updateSnapshotCount(storeID uint64, snapshotCount int)
...
+	s.testBalanceSpeed(c, testCases, 1)
+	s.testBalanceSpeed(c, testCases, 10)
+	s.testBalanceSpeed(c, testCases, 100)
+	s.testBalanceSpeed(c, testCases, 1000)
+}
+
+func (s *testBalanceSpeedSuite) testBalanceSpeed(c *C, tests []testBalanceSpeedCase, capaGB uint64) {
+	for _, t := range tests {
+		sourceCount := t.sourceCount
+		sourceRatio := float64(t.sourceCount) / float64(capaGB)
```
Comment (siddontang) [inline server/balancer_test.go]: why ratio is calculated by sourceCount / capaGB? 

capaGB is for capacity GB, it seems a size, not count.

File: server/cluster.go [hunk 12]
```diff
@@ -427,11 +428,10 @@ func (c *RaftCluster) collectMetrics() {
 	storeDownCount := 0
 	storeOfflineCount := 0
 	storeTombstoneCount := 0
-	regionTotalCount := 0
 	storageSize := uint64(0)
 	storageCapacity := uint64(0)
-	minLeaderRatio, maxLeaderRatio := float64(1.0), float64(0.0)
-	minStorageRatio, maxStorageRatio := float64(1.0), float64(0.0)
+	minLeaderScore, maxLeaderScore := math.MaxFloat64, float64(0.0)
```
Comment (siddontang) [inline server/cluster.go]: maybe confused that min is max float and max is 0.

Comment (siddontang) [inline server/cluster.go]: add a comment here.

File: server/api/store.go (hunk 10 above)
Comment (huachaohuang) [inline server/api/store.go]: No need, and it's weird to put a total count in a store too.

File: server/cluster.go (hunk 12 above)
Comment (huachaohuang) [inline server/cluster.go]: I think it is a very common way to find min and max.

**Comment by huachaohuang** [comment]
PTAL @siddontang @disksing

File: server/cluster.go [hunk 13]
```diff
@@ -451,24 +451,21 @@ func (c *RaftCluster) collectMetrics() {
...
+		storageSize += s.storageSize()
 		storageCapacity += s.stats.GetCapacity()
-		if regionTotalCount < s.stats.TotalRegionCount {
-			regionTotalCount = s.stats.TotalRegionCount
-		}
 
-		// Balance.
-		if minLeaderRatio > s.leaderRatio() {
-			minLeaderRatio = s.leaderRatio()
+		// Balance score.
```
Comment (siddontang) [inline server/cluster.go]: can we use maxFloat64 or minFloat64 like maxUint64, minUint64 directly here?

File: server/balancer.go [hunk 14]
```diff
@@ -14,16 +14,53 @@
...
 	"github.com/ngaut/log"
 	"github.com/pingcap/kvproto/pkg/metapb"
 )
 
 var storeCacheInterval = 30 * time.Second
 
+// minBalanceDiff returns the minimal diff to do balance. The formula is based
+// on experience to let the diff increase alone with the count slowly.
+func minBalanceDiff(count uint64) float64 {
+	if count < 10 {
```
Comment (siddontang) [inline server/balancer.go]: define a const var for 10 and 2

File: server/balancer_test.go [hunk 15]
```diff
@@ -216,50 +277,43 @@ func (s *testBalanceStorageSchedulerSuite) TestBalance(c *C) {
...
-	// Test MinBalanceDiffRatio.
-	// When diff storage ratio < MinBalanceDiffRatio, no schedule.
-	tc.updateRegionCount(2, 6, 0.4)
-	tc.updateRegionCount(3, 7, 0.4)
-	tc.updateRegionCount(4, 8, 0.4)
-	c.Assert(sb.Schedule(cluster), IsNil)
 }
 
-func (s *testBalanceStorageSchedulerSuite) TestReplicas3(c *C) {
+func (s *testBalanceRegionSchedulerSuite) TestReplicas3(c *C) {
```
Comment (disksing) [inline server/balancer_test.go]: Some comments should be updated in this test case. 'stroage ratio/region ratio' -> 'region count' or 'region score'.

File: server/balancer_test.go [hunk 16]
```diff
@@ -284,44 +338,44 @@ func (s *testBalanceStorageSchedulerSuite) TestReplicas3(c *C) {
...
 
-	// Store 7 has different zone with other stores but larger storage ratio than store 1.
-	tc.addLabelsStore(9, 1, 0.6, map[string]string{"zone": "z2", "rack": "r1", "host": "h1"})
+	// Store 9 has different zone with other stores but larger storage ratio than store 1.
+	tc.addLabelsStore(9, 9, map[string]string{"zone": "z2", "rack": "r1", "host": "h1"})
 	c.Assert(sb.Schedule(cluster), IsNil)
 }
 
-func (s *testBalanceStorageSchedulerSuite) TestReplicas5(c *C) {
+func (s *testBalanceRegionSchedulerSuite) TestReplicas5(c *C) {
```
Comment (disksing) [inline server/balancer_test.go]: Some comments should be updated in this test case. 'ratio' -> 'region count' or 'region score'.

File: server/util.go [hunk 17]
```diff
@@ -350,3 +350,38 @@ func GetPDMembers(etcdClient *clientv3.Client) ([]*pdpb.PDMember, error) {
...
+}
+
+func maxUint64(a, b uint64) uint64 {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+func minFloat64(a, b float64) float64 {
```
Comment (disksing) [inline server/util.go]: Why not just use `math.Min()`?

**Comment by disksing** [comment]
The rest LGTM.

**Pull request merged**